stdlib/runtime/numbers.gr

```diff
@@ -168,27 +179,59 @@ export let reducedInteger = x => {
 }
 
 @unsafe
-let reducedFraction = (x, y) => {
+let reducedBigInteger = x => {
+  if (BI.canConvertToInt64(x)) {
+    // CONVENTION: We assume that this function is called in
+    //             some sort of tail position, meaning that
+    //             the original input is no longer used after
+    //             this function returns.
+    let ret = reducedInteger(BI.toInt64(x))
+    Memory.decRef(x)
+    ret
+  } else {
+    x
+  }
+}
+
+@unsafe
+let reducedFractionBigInt = (x, y) => {
   let mut x = x
   let mut y = y
+  let mut needsDecref = false
 
-  if (WasmI32.ltS(y, 0n)) {
+  if (BI.isNegative(y)) {
     // Normalization 1: Never do negative/negative
     // Normalization 2: Never allow a negative denominator
-    x = i32neg(x)
-    y = i32neg(y)
+    needsDecref = true
+    x = BI.negate(x)
+    y = BI.negate(y)
   }
-  if (WasmI32.eqz(y)) {
+  if (BI.eqz(y)) {
     throw Exception.DivisionByZero
   }
-  if (WasmI32.eqz(WasmI32.remS(x, y))) {
-    // Avoid allocation if possible
-    reducedInteger(WasmI64.extendI32S(WasmI32.divS(x, y)))
+  let quotremResult = Memory.malloc(8n)
+  BI.quotRem(x, y, quotremResult)
+  let q = WasmI32.load(quotremResult, 0n)
+  let r = WasmI32.load(quotremResult, 4n)
+  Memory.free(quotremResult)
+  let ret = if (BI.eqz(r)) {
+    reducedBigInteger(q)
   } else {
-    // x not evenly divisible by y
-    let factor = gcd32(x, y)
-    newRational(WasmI32.divS(x, factor), WasmI32.divS(y, factor))
+    Memory.decRef(q)
+    Memory.decRef(r)
```

Comment:
Do we need decrefs here for these `WasmI32` values?

Reply:
Yes. They are malloc'ed inside of `quotRem` and do not escape here, so we need to manually manage them (they are never cast to a GC-aware type). I'll add a comment to clarify.